### ISSUE TITLE
Compiling

### ISSUE THREAD

**Issue by ihateu3** [GCodeProjects/GCodeWorkShop#14]
I really hate to be that guy, but no matter what I try, I am unable to get this to compile and work for Windows.  Is there an easier compiler, or easier way to compile this?

Thanks

**Comment by u-235** [comment]
Hi.
My main OS is Linux and so I can't give detailed advice on building EdytorNC for Windows. When I was doing test builds for Windows, I settled on [MSYS2](https://www.msys2.org).

**Comment by ihateu3** [comment]
I see that you have updated the compile instructions.  I guess I am really stupid, because as simple as they are, I still am unable to compile this.

There were many errors in QT5 when I went to run it, and so I resorted to Ubuntu, where after getting to the last step and typing "make" I got "no targets specified and no makefile found"

I really want to try this, and wouldn't even mind helping add some tooltips for Delta controllers.  Is there a way you can include  a precompiled version for windows here?

Thanks again, one day I will get it to work.

**Comment by u-235** [comment]
Hi.
This line is most likely the cause of many errors

```
qmake -r [CONFIG+=debug] [PREFIX=/usr/ | PREFIX=~/.local/bin/]
```

The options inside the square brackets are optional and can be omitted. That is, you just have to type

```
qmake -r
```

If this doesn't help, please copy the console output with errors into your message.

**Comment by ihateu3** [comment]
Ok, I will be pretty detailed to help others that have issues compiling on their journey.

The latest versions of Ubuntu do not support something (I think it was QT5?), so I had to install ubuntu-20.04.5 that does support it.
I then had to install G++
Then I was getting a serialport error when compiling so I had to install the packages.

sudo apt-get install libqt5serialport5
sudo apt-get install libqt5serialport5-dev

after going through the steps, I now can get through to "make" but the error is now 
"could not exec '/usr/lib/qt5/bin/lrelease': No such file or directory
make: *** [Makefile:287: FORCE] Error 1

But really, I need this for compiled for windows so I can use it on windows.

**Comment by u-235** [comment]
I built the application for Windows 32 bit. With all the libraries it was about 80 MB. GitHub has a limit on the size of attached files of 25 MB and only accepts zip archives. I tried to compress files directly to zip, but the archive was 30 MB. Then I compressed files to 7z and then put archive to zip.

[edytornc-x86.zip](https://github.com/u-235/edytornc/files/10472804/edytornc-x86.zip)

**Comment by ihateu3** [comment]
Thanks.  The Linux solution that you gave worked to compile it in Linux.

As for compiling in Windows, I am probably missing libraries, so if there are libraries that are needed could you add them to the windows instructions?  Thanks!

Anyways, again thanks for compiling the Windows version for me.  I see that an issue in the old versions is still present in this new build.  The "show diff" buttons are reversed.  When you click on the "open current file in the left diff window" it opens in the right window, and when you click on the  "open current file in the right diff window" it opens in the left window.
![Show diff](https://user-images.githubusercontent.com/19616711/213888372-2bda852f-eafc-498b-a777-805177a4f691.jpg)

**Comment by u-235** [comment]
I have written instructions on how to build an application using MSYS2. Could you please try to build the application and give feedback?

-----------------------------


Building EdytorNC using MSYS2
=============================

[MSYS2] provides a set of tools and libraries for developing Windows applications. New releases of MSYS2 will no longer support Windows 7 and 8.1 from 2023, so MSYS2 should be [installed][MSYS2-installation] from the archive for these operating systems.

Note that the following instructions show the build of the 64-bit version. If you need to build the 32-bit version, replace x86_64 with i686, for example, type mingw-w64-x86_64-gcc instead of mingw-w64-i686-gcc. Read more about [naming packages][MSYS2-package-naming].

One more important note: if an error occurs at some point (just an error, not a warning), further steps make sense only after the error is fixed.


Preparing the desktop environment
---------------------------------

I assume that MSYS2 is successfully installed, and its console is available to you. Next, to prepare the desktop environment you should make some simple steps, entering commands into the MSYS2 console.

### Synchronize the package manager

A good rule of thumb is to synchronize your package manager with the repositories prior to installing packages.

    pacman -Syuu


### Installing Qt

We install the base package Qt5 and its translations, Qt tools, and network and serial port libraries.

    pacman -S mingw-w64-x86_64-qt5-base mingw-w64-x86_64-qt5-serialport mingw-w64-x86_64-qt5-tools mingw-w64-x86_64-qt5-translations


### Installing GCC

To build an application we need a C++ compiler and a pair of utilities.

    pacman -S mingw-w64-x86_64-gcc make unzip

You are now ready to build the application.


### Building the application

We download the archive with the current version of the `master` branch into our home directory, unzip it, and jump to the resulting project directory. Then we run `qmake` and `make` to build the application.

    cd ~
    wget https://github.com/u-235/edytornc/archive/refs/heads/master.zip
    unzip master.zip
    cd edytornc-master
    qmake -r
    make -j4

After completing the build, a new bin directory with the application will appear in the project directory.

### Running the application

So, you are in the `~/edytornc-master` directory and it's time to test run the application. After running the command

    bin/release/edytornc.exe

the application window should appear and you can be congratulated on successfully building the application. However, if you try to start the application from an explorer, you will get a warning that some dlls required for the application are missing. To fix this problem, we are going to copy the application and all necessary files into `C:/EdytorNC`. To do so we launch the script

    install/MSYS2-deploy.sh

The application is now available at `C:/EdytorNC/bin/edytornc.exe` and a link can be created on the desktop for easy launch.


[MSYS2]: https://www.msys2.org/
[MSYS2-installation]: https://www.msys2.org/wiki/MSYS2-installation/
[MSYS2-package-naming]: https://www.msys2.org/docs/package-naming/

**Comment by ihateu3** [comment]
I followed all the steps in order, but got an error when I went to run make -r



<img width="350" alt="image" src="https://user-images.githubusercontent.com/19616711/215611204-9e64840d-b208-45dc-bacf-aa6bb96c68fe.png">

**Comment by u-235** [comment]
I forgot that `MSYS2` can be run in different configurations. You run it in `UCRT64` mode, but you need `MINGW64` mode.

**Comment by ihateu3** [comment]
Can you update the instructions on how to do that?  I am a complete noob when it comes to compiling.

**Comment by u-235** [comment]
There are several shortcuts in the Start menu for `MSYS2`. It is necessary to click `MINGW64`.

**Comment by ihateu3** [comment]
Ok, so now I get a few files it cannot find after running the "qmake -r".  I did go ahead and try to run "make -j4" which had a bunch of warnings and one error.

I then tried to "bin/release/edytornc.exe" and it said there was no such file.

![image](https://user-images.githubusercontent.com/19616711/215889476-1b676e3a-4bd7-4cf8-8f51-3ea38a8f7f56.png)

<img width="581" alt="image" src="https://user-images.githubusercontent.com/19616711/215889675-9f65b821-0411-41dc-ac9d-5af6076352e1.png">

**Comment by u-235** [comment]
Please, run `make` again and show an error.

**Comment by ihateu3** [comment]
I ran qmake -r again and get the same error as in the pictures I attached.  

WARNING: Failure to find: edytornc_res.o
WARNING: Failure to find: sfs_res.o

**Comment by u-235** [comment]
This is an error due to a peculiarity of the GCC compiler. It is strange that I have not had such an error. But in any case it should be fixed.

**Comment by u-235** [comment]
I've fixed some bugs and updated the master branch, so you can try to build the application. You need to remove the old files, to do this run the following commands in `MSYS2`.

```
cd ~
rm master.zip
rm -r -f edytornc-master
```

Then repeat the build from the "Building the application" phase.

**Comment by ihateu3** [comment]
Perfect, it works!!!!

The about section doesn't show the version though that I can see?  Might be a separate issue, I just noticed that.  Regardless, I can now compile it!  And also the buttons are now correct!

**Comment by u-235** [comment]
Congratulations on your new skill! Apart from clarifying about choosing the correct `MSYS2` startup mode, is there anything I need to add or change in the build instructions? I plan to make instructions with general guidelines for all platforms and separately for `MSYS2`.

The application version is usually changed at release, which implies significant changes to the software. In my opinion, there are no such changes yet.

**Comment by ihateu3** [comment]
I do believe that these instructions are good.  I had no issues once the correct MSYS2 was selected, and the code updated to fix the errors I was having.

**Comment by u-235** [comment]
In that case, I will close this issue when I add the documentation. Thank you for the detailed feedback.

**Comment by WD50** [comment]
finally compiled, and don't forget to open MINGW64 with admin rights

**Comment by u-235** [comment]
Why open MINGW64 with admin rights?

**Comment by WD50** [comment]
The new Windows 10/11 are very rigorous where all programs can write something, especially if it writes directly to C:
So a right click and run as administrator on MINGW64 is simply to ensure that there will be no problems with that.

**Comment by u-235** [comment]
The installer for Windows is now available.

https://github.com/GCodeProjects/GCodeWorkShop/releases/tag/2024-05
